### PR TITLE
Add an example for using open_instruct data

### PR DESCRIPTION
**Pull request by vwxyzjn**
Hi @epwalsh and @soldni, here is an example to create fine-tuning data with the open instruct utility.

I am a bit unsure about several things:

1. how to integrate attention mask / labels in olmo-core
2. where you'd like to store this `data_transformation.py`

Instructions on how to use 

```bash
python src/examples/llama/oi_sft_data.py
torchrun --nproc-per-node=1 src/examples/llama/train_sft.py haha
```

image

### CODE REVIEW

File: src/oi_data/data_transformation.py
Comment (epwalsh) [inline src/oi_data/data_transformation.py:1]: I think this could live in `src/olmo_core/data/` as a submodule there, and then we just add `datasets` and `huggingface-hub` as additional optional dependencies (I think those are the only new ones) to `pyproject.toml`.

File: src/olmo_core/data/oi_data/data_transformation.py [hunk 1]
```diff
@@ -0,0 +1,1531 @@
+# this file deals with dataset pre-processing before training
```
Comment (tyler-romero) [inline src/olmo_core/data/oi_data/data_transformation.py]: note: exact copy from open-instruct

**Comment by tyler-romero** [comment]
Script is pretty much ready for integration testing now!

Realistic command to launch SFT:
```
uv run python src/scripts/train/sft/OLMo2-7B-sft.py launch olmo2-7B-sft-take3-8gpu /weka/oe-training-default/ai2-llm/checkpoints/dustins/lc_7b_cont_pretrain_final_anneal/step11921 ai2/augusta-google-1 --trainer.callbacks.wandb.enabled=True --launch.num_gpus=8 --launch.priority=high
```

File: src/olmo_core/train/trainer.py [hunk 2]
```diff
@@ -743,10 +743,19 @@ def load_checkpoint(self, dir: PathOrStr, *, load_trainer_state: Optional[bool]
...
+        error: Optional[Exception] = None
         if get_rank() == 0 and not self.checkpointer.dir_is_checkpoint(dir):
-            # Try to find the latest checkpoint in the directory.
-            dir = self.checkpointer.latest_checkpoint(dir)
-        dir = scatter_object(dir)
+            try:
+                dir_to_scatter = self.checkpointer.latest_checkpoint(dir)
+            except FileNotFoundError as e:  # defer raising until after the scatter
+                dir_to_scatter, error = None, e
+        dir_to_scatter = scatter_object(dir_to_scatter)
```
Comment (tyler-romero) [inline src/olmo_core/train/trainer.py:765]: This avoids a 30 min hang (rank 0 throws exception while other ranks are waiting) when the checkpoint isn't found

**Review by natolambert — APPROVED** [review]
Did a quick skim, generally looks fine, some TODOs and placeholders. how is the bos token before a sequence handled in olmo_core? One of the flags we need in open instruct is --add_bos for olmo's

File: src/olmo_core/train/trainer.py (hunk 2 above)
Comment (epwalsh) [reply src/olmo_core/train/trainer.py:765]: Oh, this is a great fix!

File: src/scripts/train/sft/OLMo2-7B-sft.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+            .with_callback("gpu_monitor", GPUMemoryMonitorCallback())
+            .with_callback("config_saver", ConfigSaverCallback())
+            .with_callback("garbage_collector", GarbageCollectorCallback())
+            .with_callback(
+                "checkpointer",
+                CheckpointerCallback(
+                    save_interval=1000, ephemeral_save_interval=500, save_async=True
+                ),
+            )
+            .with_callback(
```
Comment (epwalsh) [inline src/scripts/train/sft/OLMo2-7B-sft.py]: Consider using `.with_recommended_evals()` instead.
https://github.com/allenai/OLMo-core/blob/796c2dc2b57985b160395b29855bc4320ac4eea1/src/olmo_core/train/config.py#L69

**Review by epwalsh — COMMENTED** [review]
Overall looks pretty good, only a couple minor comments. Also, have we ablated document packing + intra-document masking in this scenario yet? I suspect it works well, but we should confirm if we haven't already.

**Comment by tyler-romero** [comment]
@natolambert bos is supported in olmo-core, since your review I added better support for reading the same tokenizer setup that was used to tokenize the SFT data.

**Comment by tyler-romero** [comment]
> Also, have we ablated document packing + intra-document masking in this scenario yet? I suspect it works well, but we should confirm if we haven't already.

Running that ablation overnight!

**Comment by tyler-romero** [comment]
> how is the bos token before a sequence handled in olmo_core?

bos is supported explicitly, just need to ensure that the tokenizer that is specified is configured with the right token id.
